### PR TITLE
[CodeGen] Add a pattern to fold extract_slice consumer into xfer.write.

### PR DESCRIPTION
**Pull request by hanhanW**
The LLVMCPU and SPIRV lit tests are updated because the pass runs `scf::ForOp` canonicalization patterns. It drops the unused for op results which makes analysis easier.

This is helpful for bufferization when masking is not involved. Because we make the chain simpler and the new tensor.empty op can be replaced with `flow.tensor.load` ops in bufferization pre-processing passes.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/vectorize_with_masking_and_hoist.mlir [hunk 1]
```diff
@@ -22,16 +22,16 @@
 // CHECK-NEXT:      %[[OUT_SLICE:.*]] = tensor.extract_slice %[[OUT_TENSOR_1]]{{.*}} : tensor<1024x1024xf32> to tensor<8x?xf32>
 // CHECK-NEXT:      %[[OUT_SLICE_1:.*]] = tensor.extract_slice %[[OUT_SLICE]]{{.*}} : tensor<8x?xf32> to tensor<8x?xf32>
 // CHECK-NEXT:      %[[OUT_VEC:.*]] = vector.transfer_read %[[OUT_TENSOR_1]]{{.*}} : tensor<1024x1024xf32>, vector<8x[16]xf32>
-// CHECK-NEXT:      %[[INNER_LOOP:.*]]:3 = scf.for {{.*}} iter_args({{.*}}, %[[RES:.*]] = %[[OUT_VEC]]) -> (tensor<8x?xf32>, tensor<8x?xf32>, vector<8x[16]xf32>) {
+// CHECK-NEXT:      %[[INNER_LOOP:.*]] = scf.for {{.*}} iter_args(%[[RES:.*]] = %[[OUT_VEC]]) -> (vector<8x[16]xf32>) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/vectorize_with_masking_and_hoist.mlir:25]: @banach-space FYI that I updated the test you added. I need to run the canonicalization patterns from `scf.for` and `vector.transfer_write` to make the result of hoisting simpler. It should be valid because it only drops unused results from `scf.for` op.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 2]
```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
   }
 };
 
+/// Fold tensor.extract_slice into vector.transfer_write if
+///   1. The tensor.extract_slice op has only one use.
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: The pattern looks like it is checking that the `transfer_write` has a single use (the source for the `extract_slice`).
```suggestion
///   1. The vector.transfer_write op has only one use.
```

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 3]
```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
...
+    }
+    if (!result.hasOneUse()) {
+      return failure();
+    }
+    if (!llvm::all_of(extractSliceOp.getMixedOffsets(), isZeroIndex)) {
+      return failure();
+    }
+
+    auto xferOp =
+        extractSliceOp.getSource().getDefiningOp<vector::TransferWriteOp>();
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:247]: You can move the `hasOneUse` check to the xferOp here, and then you don't need to check that `extractSliceOp.getSource()` is an `OpResult` anymore.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 4]
```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
...
+    }
+    if (!xferOp.getSource().getDefiningOp<tensor::EmptyOp>()) {
+      return failure();
+    }
+    if (xferOp.getMask()) {
+      return failure();
+    }
+    if (!xferOp.getInBounds()) {
+      return failure();
+    }
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: The default for `InBounds` is all false right? The below logic looks to conservatively handle that fine to me, is this check needed?

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 5]
```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
...
+
 } // namespace
 
 void mlir::iree_compiler::populateVectorTransferTensorSliceTransforms(
     RewritePatternSet &patterns, PatternBenefit benefit) {
-  patterns
-      .add<FoldExtractSliceIntoTransferRead, FoldInsertSliceIntoTransferWrite>(
-          patterns.getContext(), benefit);
+  patterns.add<FoldExtractSliceIntoTransferRead,
+               FoldInsertSliceIntoTransferWrite, FoldExtractSliceIntoXferWrite>(
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: Can we call it `FoldExtractSliceIntoTransferWrite` to be consistent with the other patterns here?

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: good catch!

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 4 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: Is it? I thought the in_bounds is an optional attribute. And the verifier accepts that it is empty.

See 

https://github.com/llvm/llvm-project/blob/6e6da74c8b936e457ca5e56a828823ae6a9f9066/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L4327-L4330

https://github.com/llvm/llvm-project/blob/6e6da74c8b936e457ca5e56a828823ae6a9f9066/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L3813-L3826

https://github.com/llvm/llvm-project/blob/78eb2c2cba5ad0f9a36ec1959371527313a76bbe/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L1523-L1531

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 5 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: SG, done.

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 4 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: Right, what I meant was that when `in_bounds` is empty, that is equivalent to `in_bound = [false, ..., false]`, so there is no need for this check because you're already handling the all false case.

File: compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+// CHECK-LABEL: func.func @batch_matmul_with_padding_strategy
+// CHECK-DAG:     %[[C0:.+]] = arith.constant 0 : index
+// CHECK:         %[[SCF:.+]] = scf.for {{.+}} -> (vector<1x64x128xf16>) {
+// CHECK:         }
+// CHECK:         %[[INIT:.+]] = tensor.empty(%{{.+}}) : tensor<1x?x128xf16>
+// CHECK:         %[[WRITE:.+]] = vector.transfer_write %[[SCF]], %[[INIT]]
+// CHECK-SAME:      [%[[C0]], %[[C0]], %[[C0]]] {in_bounds = [true, false, true]}
+// CHECK-SAME:      : vector<1x64x128xf16>, tensor<1x?x128xf16>
+// CHECK:         flow.dispatch.tensor.store %[[WRITE]]
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/test/optimize_tensor_insert_extract_slices.mlir:67]: Can you add a test that just checks this pattern! It is hard to follow how this pattern is getting applied for this test because it looks like it is dependent on tensor subset hoisting patterns also.

**Review by qedawkins — APPROVED** [review]
LGTM!

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp (hunk 4 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp]: I see, makes sense to me now. Thank you!

File: compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp [hunk 7]
```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
...
+///   : vector<1x64x128xf16>, tensor<1x64x128xf16>
+/// %extracted_slice = tensor.extract_slice %0[0, 0, 0] [1, %3, 128] [1, 1, 1]
+///   : tensor<1x64x128xf16> to tensor<1x?x128xf16>
+/// ```
+/// is rewritten to:
+/// ```
+/// %1 = vector.transfer_write %v, %t2[%a, %b, %c]
+///   {in_bounds = [true, false, true]}
+///   : vector<4x5xf32>, tensor<?x?xf32>
+/// ```
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:228]: Looking at dims in this example - are these correct? For example, `vector<1x64x128xf16>` vs `vector<4x5xf32>`?

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp:228]: oh, I did not update the types... good catch!
